Orbis 1.19 overhaul

kruthers opened the pull request:
This PR will overhall all of orbis to use world generation instead of functions. This means that Orbis will only effect new chunks now. This RP will be updated with changes as they happen.

### Changes
- Cooler caves now respects biome boundries
- Some towers will spawn in diffrent locations now due to limitations with custom world generation.
- Some towers/dungeons will spawn in diffrent biomes before.
- All loot tables have been overhalled

### Dangerous Dungeons Overhall
- Renamed default dungeon type to stone
- All new dungeons make use of template pools to change up spawner/ chest layouts and wall location
- Added 4 new stone type dungeons
- Added 4 new desert type dungeons
- Added 4 new badlands type dungeons
- Added 4 new ocean type dungeons
- Added a new dungeon type: `Deep slate`
- Added 4 new deep slate type dungeons
- Added a new dungeon type: `Lush`
- Added 4 new lush type dungeons
- Overhalled dungeon placement in worlds
- Deepslate and lush dungeons spawn in there own structure set to allow for them to be seperate

### Tower Structures overhall
- Added 2 new ship types
- All ships now have spawners/ chests
- All ships use template pools for sails and chests/ spawners
- Added a new spawner set themed arround pirates

### Planned Features

- [x] Port cooler caves to be modifed versions of default biomes (Done by @misode )
- [x] Updated all dungons to be structures that can spawn underground.
- [x] Updated all towers to be structures that are placed during world gen.
- [x] All towers and dungeons that have use noise in the structure now have noise applied by the template pool when its place. This creates more randomization and means things like cobwebs are also randomised within structures.
- [x] Add new tower types that use jigsaws to have diffrent tops/bottoms/middles
- [x] Add more dungeons that are actually dangrous (Will also use jigswas)
- [x] Overhall loottables in dungeons and towers
- [x] Look at maybe adding more spawner types/ groups

TheThanathor commented:
Issues I've found (most of which are to be fixed by me soon)

- large boat mast does not connect to flag
- large boat door is open
- stone dungeon does not connect walls properly 
- most stone dungeons do not properly connect walls to the ceiling
- desert dungeon has sand floating
- desert dungeon didn't have yellow concrete powder replaced
- spider deepslate dungeons spawns mobs other than spiders
- tall deepslate dungeon spawner is not connected to chain

TheThanathor commented:
- some build problems, they've been fixed and saved, need to be re-added
- desert medium 1 does not generate the extra feature
- spider spawners (gm4_orbis:spawner/spider) spawn spawners other than spiders
- ship loot fills up the chest too much and is too good, I found chests with multiple gold blocks and diamonds.
- dungeon loot is good sometimes, but other times it doesn't come with anything "valuable" like diamond items, enchanted books, saddle, music disc or golden apples, instead having just an iron shovel. Items should also be lightly enchanted more often.

TheThanathor commented on gm4_dangerous_dungeons/data/gm4_dangerous_dungeons/worldgen/structure_set/dungeons_upper.json:
```diff
@@ -0,0 +1,38 @@
+      "structure": "gm4_dangerous_dungeons:desert",
+      "weight": 3
+    },
+    {
+      "structure": "gm4_dangerous_dungeons:ocean",
+      "weight": 3
+    },
```
This file doesn't exist

kruthers replied:
Good catch, i forgot to update it in there when i split them

TheThanathor commented:
- [ ] Tower loot tables have a chance to have 2 music discs in one chest
- [ ] Tower loot table gear should be low level enchanted more often
- [ ] Old dungeons spawn without loot in chests (might be because I didn't have `gm4_orbis`)
- [ ] Dispenser traps are empty, loot table `gm4_orbis:traps/arrow_dispenser`, should have a few arrows
- [ ] Desert medium 1 doesn't generate extra feature
- [ ] `gm4_dangerous_dungeons:dungeon/stone/medium_2/decor/lower_floor/c` has been resaved and needs to be re-added

TheThanathor commented:
- [ ] towers were resaved and need to be readded
- [ ] new stone dungeons should generate under oceans just like under plains since ocean dungeons are now above ground
- [ ] ocean dungeons should not spawn above y55 (or similar below sea level if possible)
- [ ] `gm4_tower_structures:towers/ocean/large_ship/features/3` has been resaved
- [ ] arrow traps still have no arrows (`gm4_orbis:traps/arrow_dispenser`)
- [ ] `gm4_dangerous_dungeons:dungeon/badlands/medium_1/base` has been resaved
- [ ] desert medium 1 :sob:

TheThanathor commented:
- skeletons in towers and dungeons spawn without bows
- mobs in towers do not spawn in skylight
- tower spawn rate should be about halved
- ship spawn rate should be cut by four
- spawn rate ocean dungeons should be halved
- deepslate/lush seems about right, maybe increase deepslate by like 25%
- stone/desert/badlands can be a bit more common, like 50-75% more

kruthers commented:
Bug from public server: Spawners look to have the wrong data for light requirments

misode commented on gm4_tower_structures/data/gm4_guidebook/advancements/tower_structures/page_0.json:
```diff
@@ -17,239 +17,230 @@
-      "trigger": "minecraft:player_generates_container_loot",
+    "badlands": {
+      "trigger": "minecraft:location",
       "conditions": {
```
This file is missing the necessary guidebook `load.status` score checks, which causes the unlock page message to be displayed in chat even when the guidebook module is not enabled.

SpecialBuilder32 commented:
@kruthers I have a design question that arose while investigating #837. The spawners created by orbis all have explicit light-level rules given by `custom_spawn_rules`, which appear to make the spawners behave like vanilla spawners before 1.19, spawning monsters at higher light levels. Whats the reason for this change

Incidentally, this is what causes orbis spawners to not be accepted by spawner minecarts at the moment, though that may be a change required by the spawner minecarts module.

EpyonProjects commented:
Dangerous Dungeons and Tower Structures guidebook unlocks do not check for Guidebook being loaded.  Cooler Caves does.

Bloo-dev commented:
> @kruthers I have a design question that arose while investigating #837. The spawners created by orbis all have explicit light-level rules given by `custom_spawn_rules`, which appear to make the spawners behave like vanilla spawners before 1.19, spawning monsters at higher light levels. Whats the reason for this change

@SpecialBuilder32 I have gone through and removed the artificial light level limitations, as they were causing some other issues.
I have also cleaned up the spawner NBT data as some of the spawner had quite some duplicate NBT.

kruthers commented:
> > @kruthers I have a design question that arose while investigating #837. The spawners created by orbis all have explicit light-level rules given by `custom_spawn_rules`, which appear to make the spawners behave like vanilla spawners before 1.19, spawning monsters at higher light levels. Whats the reason for this change
> 
> @SpecialBuilder32 I have gone through and removed the artificial light level limitations, as they were causing some other issues. I have also cleaned up the spawner NBT data as some of the spawner had quite some duplicate NBT.

The light levels were there to try and allow spawners to work in towers, as default limitations stopped tower spawners from working due to light level. Although i belive i may have not set them up fully correctly

SpecialBuilder32 commented on gm4_cooler_caves/data/minecraft/worldgen/noise_settings/overworld.json:
I'm confused about the purpose of this file, as it doesn't reference any gamemode 4 content. Isn't the purpose of biome_extensions to remove the need to overwrite vanilla worldgen files such as this? Is this just in the wrong place?

BluePsychoRanger replied:
lib_biome_extensions only provides compatibility with placed features in biomes. The overworld noise settings (surface rules) cannot be made compatible by a library and must be overwritten. This is used for the badlands stone to terracotta, desert stone to sand stone and, and snowy stone to snow blocks.
